consul/autopilot.go

```diff
@@ -0,0 +1,286 @@
+	for {
+		select {
+		case <-s.autopilotShutdownCh:
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			if err := s.pruneDeadServers(); err != nil {
+				s.logger.Printf("[ERR] consul: error checking for dead servers to remove: %s", err)
+			}
+		case <-s.autopilotRemoveDeadCh:
+			if err := s.pruneDeadServers(); err != nil {
+				s.logger.Printf("[ERR] consul: error checking for dead servers to remove: %s", err)
+			}
+		}
+	}
+}
+
+// pruneDeadServers removes up to numPeers/2 failed servers
+func (s *Server) pruneDeadServers() error {
+	state := s.fsm.State()
+	_, autopilotConf, err := state.AutopilotConfig()
+	if err != nil {
+		return err
+	}
+
+	// Look for dead servers to clean up
+	peers, err := s.numPeers()
+	if err != nil {
+		return err
+	}
+
+	removed := 0
+	if autopilotConf.CleanupDeadServers {
+		for _, member := range s.serfLAN.Members() {
+			// Exit early if we already removed the max amount of servers
+			if removed == peers/2 {
+				break
+			}
+
+			valid, _ := agent.IsConsulServer(member)
+			if valid && member.Status == serf.StatusFailed {
+				removed++
+				s.logger.Printf("[INFO] consul: Attempting removal of failed server: %v", member.Name)
+				go s.serfLAN.RemoveFailedNode(member.Name)
+			}
+		}
+	}
+
+	return nil
+}
+
+// promoteNonVoters promotes eligible non-voting servers to voters.
+func (s *Server) promoteNonVoters(autopilotConf *structs.AutopilotConfig) error {
+	minRaftProtocol, err := ServerMinRaftProtocol(s.LANMembers())
+	if err != nil {
+		return fmt.Errorf("error getting server raft protocol versions: %s", err)
+	}
+
+	if minRaftProtocol >= 3 {
+		future := s.raft.GetConfiguration()
+		if err := future.Error(); err != nil {
+			return fmt.Errorf("failed to get raft configuration: %v", err)
+		}
+
+		var promotions []raft.Server
+		raftServers := future.Configuration().Servers
+		voterCount := 0
+		for _, server := range raftServers {
+			// If this server has been stable and passing for long enough, promote it to a voter
+			if server.Suffrage == raft.Nonvoter {
+				health := s.getServerHealth(string(server.Address))
+				if health != nil && health.Healthy && time.Now().Sub(health.StableSince) >= autopilotConf.ServerStabilizationTime {
+					promotions = append(promotions, server)
+				}
+			} else {
+				voterCount++
+			}
+		}
+
+		// Exit early if there's nothing to promote
+		if len(promotions) == 0 {
+			return nil
+		}
+
+		// If there's currently an even number of servers, we can promote the first server in the list
+		// to get to an odd-sized quorum
+		newServers := false
+		if voterCount%2 == 0 {
+			addFuture := s.raft.AddVoter(promotions[0].ID, promotions[0].Address, 0, 0)
+			if err := addFuture.Error(); err != nil {
+				return fmt.Errorf("failed to add raft peer: %v", err)
+			}
+			promotions = promotions[1:]
+			newServers = true
+		}
+
+		// Promote remaining servers in twos to maintain an odd quorum size
+		for i := 0; i < len(promotions)-1; i += 2 {
```

Comment:
if there's only one remaining in promotions at this point, it won't enter the loop at all